ruby/phase-1.md

```diff
@@ -0,0 +1,536 @@
+in memory and manage accordingly to the worker configuration.
+
+Here's a diagram to demonstrate the whole setup:
+
+<img src="./ruby_threads.png" width="400" align="center" alt="Temporal" />
+
+### Alternatives considered
+
+We've decided against using the `Client` to initialise the `Worker` (something that Go and Python
+SDKs are doing). It seemed like a wrong abstraction and would require making many attributes public
+on the client to make sure the worker can access them (expanding the public API of a client). Using
+a configuration object solves this and opens up further optimisation opportunities (that will be
+proposed in the 2nd phase).
+
+We can potentially use the reactor pattern via Async gem. This has the benefit of giving extra
+performance, however it would drastically reduce the performance unless activities are written in a
+reactor-friendly way. This can be further added as an optional approach to apps already using Async,
+but doesn't feel like a great default option.
+
+Additionally the SDK users will want to leverage multiple CPUs by spawning multiple worker
+processes. We have agreed that this part should be left out of the Ruby SDK and handled by
+developers themselves. Potentially this functionality can be added as a separate Ruby gem.
+
+
+## Activities
+
+Each activity is going to be represented by its own class. This decision comes with a few benefits:
+
+- There's no need to pollute global scope with activity methods (see "Global Scope" section)
+- There's no need to inject global context for communicating with Temporal. It will be available
+  through a private variable on the class instance (`#activity`)
+- It plays nicely with Ruby's OOP model
+- It allows for future extensibility including 3rd party extensions
+- Activity can be easily tested by injecting a test context into the class instead of the real one
+
+By default we'll use the name of the class as activity name (when communicating with the server).
+This is of course overridable in case the SDK users would like to use their own naming schema. And
+we'll use the method `#execute` as the primary entrypoint for executing user code.
+
+```ruby
+class MyActivity < Temporal::Activity
+  ManuallyCancelled = Class.new(Temporal::Error)
+
+  def execute(arg_1, arg_2: nil)
+    loop do
+      sleep 1
+      activity.logger.info('❤️')
+      activity.logger.debug("Attempt #{activity.info.attempts}")
+      activity.heartbeat('details')
+    end
+  rescue Temporal::Error::ActivityCancelled
+    raise ManuallyCancelled, 'activity got cancelled'
+  end
+end
+```
+
+Notes:
+
+- `activity` is an instance of `Temporal::Activity::Context`
+- `activity.logger` will use a logging adapter configure via `Temporal::Configuration` (more on
+  this in the 2nd phase)
+- Common logging context can be configured by using middleware (more on this in the 2nd phase)
+- `activity.heartbeat` is a synchronous call
```

Comment:
In Core based SDKs heartbeat is an internal call, you will not get notified of cancellation via this mechanism.
I'm assuming you did not have this context when working on the proposal.

Reply:
I did not, no. How does it notify of a cancellation then? Where can I find more info on this?

Reply:
You will get notified of cancellation when the worker polls on an activity task: https://github.com/temporalio/sdk-core/blob/6dd3a1a687386e81792b16f2b579b824f1529780/protos/local/temporal/sdk/core/activity_task/activity_task.proto#L12

It's up to you to decide the best way to surface cancellation in the activity.

Reply:
Oh… but what if a cancellation came in during a long-running activity? Surely you don't expect an activity task for an activity that's being processed right now. This is also similar to how Python does it — https://github.com/temporalio/proposals/blob/master/python/phase-1.md?plain=1#L348. What am I missing?